sycl/doc/extensions/proposed/sycl_ext_oneapi_raw_kernel.asciidoc

```diff
@@ -0,0 +1,513 @@
+* link:../proposed/sycl_ext_oneapi_free_function_queries.asciidoc[
+  sycl_ext_oneapi_free_function_queries]
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.
+Interfaces defined in this specification may not be implemented yet or may be
+in a preliminary state.
+The specification itself may also change in incompatible ways before it is
+finalized.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Overview
+
+This extension introduces a new way to define a kernel as a simple C++
+function, where the kernel arguments are parameters to the function.
+This is different from standard SYCL kernels, where the kernel arguments are
+either captures of a lambda expression or member variables of a callable
+object.
+
+The primary motivation for this extension is the
+link:../proposed/sycl_ext_oneapi_kernel_compiler.asciidoc[
+sycl_ext_oneapi_kernel_compiler], which allows online compilation of a kernel
+from source code.
+These kernels must have a clean separation between host and device code, and
+they must have a defined order to their arguments because the application sets
+the argument values by their "index" via `handler::set_arg(index, value)`.
+Because variables captured by lambda expressions have no defined order, we need
+some other way to define kernel arguments for use with the
+sycl_ext_oneapi_kernel_compiler extension, and the sycl_ext_oneapi_raw_kernel
+extension provides that mechanism.
+The "raw kernel" feature is a separate extension, though, because it can also
+be used independently from sycl_ext_oneapi_kernel_compiler.
+For example, some users may find it more familiar to define kernels as plain
+functions or they might have other reasons to prefer a clean separation between
+host and device code.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_RAW_KERNEL`
+to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Defining a raw kernel
+
+A raw kernel is a normal C++ function definition, where the function definition
+is decorated with one of the following compile-time properties: `range_kernel`,
+`nd_range_kernel`, or `single_task_kernel`.
+Each of these properties takes a template parameter `Name`, which serves as the
+name of the kernel, similar to the `KernelName` template parameter to
+`parallel_for` for a standard SYCL kernel.
```

Comment:
Why not expose this as a separate property?  If we did, we could consider removing the `KernelName` template parameter from the `sycl::parallel_for` interface (in the enqueue extension) and have users provide it via a property instead.

Reply:
+1.
User should be able to get access to index space identifiers via free function calls. There should be no need to have compile time kernel properties.

Reply:
The latest proposal removes the concept of using a type-name to identify the kernel (a7bef5ed34e9e71199da439fa5d27015782c191d).  Therefore, I think this conversation is moot, so I'm marking it resolved.